File: go/grimlock/types/v1.go
package types

// V1 version-tagged types for type safety

// KeyPairV1 represents a v1 X25519 key pair with version tag
type KeyPairV1 struct {
	KeyPair
	Version string `json:"_version"`
}

// EncryptedPrivateKeyV1 represents a v1 encrypted private key with version tag
type EncryptedPrivateKeyV1 struct {
	EncryptedPrivateKey
	Version string `json:"_version"`
}

// EncryptedMessageV1 represents a v1 encrypted message with version tag
type EncryptedMessageV1 struct {
	EncryptedMessage
	Version string `json:"_version"`
}

// MessagePayloadV1 represents a v1 message payload with version tag
type MessagePayloadV1 struct {
	MessagePayload
	Version string `json:"_version"`
}

// NewKeyPairV1 creates a new v1 key pair with version tag
func NewKeyPairV1(privateKey, publicKey []byte) *KeyPairV1 {
	return &KeyPairV1{
		KeyPair: KeyPair{
			PrivateKey: privateKey,
			PublicKey:  publicKey,
		},
		Version: "v1",
	}
}

// NewEncryptedPrivateKeyV1 creates a new v1 encrypted private key with version tag
func NewEncryptedPrivateKeyV1(iv, tag, ciphertext []byte) *EncryptedPrivateKeyV1 {
	return &EncryptedPrivateKeyV1{
		EncryptedPrivateKey: EncryptedPrivateKey{
			IV:         iv,
			Tag:        tag,
			Ciphertext: ciphertext,
		},
		Version: "v1",
	}
}

// NewEncryptedMessageV1 creates a new v1 encrypted message with version tag
func NewEncryptedMessageV1(ephPubKey, iv, tag, ciphertext []byte) *EncryptedMessageV1 {
	return &EncryptedMessageV1{
		EncryptedMessage: EncryptedMessage{
			EphemeralPublicKey: ephPubKey,
			IV:                 iv,
			Tag:                tag,
			Ciphertext:         ciphertext,
		},
		Version: "v1",
	}
}

File: python/grimlock/types/v1.py
"""V1-specific type definitions for Grimlock crypto module."""

from typing import Literal

from .common import EncryptedMessage, EncryptedPrivateKey, KeyPair


class KeyPairV1(KeyPair):
    """V1 key pair with version tag."""

    _version: Literal["v1"] = "v1"


class EncryptedMessageV1(EncryptedMessage):
    """V1 encrypted message with version tag."""

    _version: Literal["v1"] = "v1"


class EncryptedPrivateKeyV1(EncryptedPrivateKey):
    """V1 encrypted private key with version tag."""

    _version: Literal["v1"] = "v1"


def new_key_pair_v1(private_key: bytes, public_key: bytes) -> KeyPairV1:
    """Create a new v1 key pair with version tag."""
    return KeyPairV1(private_key=private_key, public_key=public_key)


def new_encrypted_private_key_v1(
    iv: bytes, tag: bytes, ciphertext: bytes
) -> EncryptedPrivateKeyV1:
    """Create a new v1 encrypted private key with version tag."""
    return EncryptedPrivateKeyV1(iv=iv, tag=tag, ciphertext=ciphertext)


def new_encrypted_message_v1(
    eph_pub_key: bytes, iv: bytes, tag: bytes, ciphertext: bytes
) -> EncryptedMessageV1:
    """Create a new v1 encrypted message with version tag."""
    return EncryptedMessageV1(
        ephemeral_public_key=eph_pub_key,
        iv=iv,
        tag=tag,
        ciphertext=ciphertext,
    )
